resources/ner_text_extraction_pipeline/pipeline.py

```diff
@@ -0,0 +1,237 @@
+        is_split_into_words=True,
+        return_offsets_mapping=True,
+        padding="max_length",
+        truncation=True,
+        max_length=MAX_NER_INPUT_LEN,
+        return_tensors="pt",
+    )
+
+
+def ner_inference(model, preprocessed_input):
+    """
+    Perform the NER inference.
+    Copied from https://github.com/ror-community/affiliation-matching-experimental/tree/main/ner_tests/inference
+
+    :param model: NER model object
+    :param preprocessed_input: Preprocessed input object
+    :return: NER token predictions
+    """
+
+    ids = preprocessed_input["input_ids"]
+    mask = preprocessed_input["attention_mask"]
+    outputs = model(ids, attention_mask=mask)
+    logits = outputs[0]
+    active_logits = logits.view(-1, model.num_labels)
+    predictions = torch.argmax(active_logits, axis=1)
+    return predictions
+
+
+def postprocess_output(tokenizer, ner_input, token_predictions):
+    """
+    Postprocess NER inference output
+    Adapted from https://github.com/ror-community/affiliation-matching-experimental/tree/main/ner_tests/inference
+
+    :param tokenizer: NER tokenizer object
+    :param ner_input: Preprocessed input object
+    :param token_predictions: NER raw predictions
+    :return: Token prediction labels or None if no organisation was detected
+    """
+
+    ids_to_labels = {0: "B-ORG", 1: "I-ORG", 2: "O", 3: "B-LOC", 4: "I-LOC"}
+    token_labels = [ids_to_labels[i] for i in token_predictions.cpu().numpy()]
+    predictions = []
+    for token_label, mapping in zip(
+        token_labels, ner_input["offset_mapping"].squeeze().tolist()
+    ):
+        if mapping[0] == 0 and mapping[1] != 0:
+            predictions.append(token_label)
+    if (
+        all(label in ["I-ORG", "B-LOC", "I-LOC", "O"] for label in predictions)
+        and "B-ORG" not in predictions
+    ):
+        return None
+    return predictions
+
+
+def extract_organisation_names(tokenizer, model, text):
+    """
+    Extract organisation names from text using a NER model.
+    Adapted from https://github.com/ror-community/affiliation-matching-experimental/tree/main/ner_tests/inference
+
+    :param tokenizer: NER tokenizer object
+    :param model: NER model object
+    :param text: Input text
+    :return: List of extracted organisation names
+    """
+
+    tokens = text.lower().split()
+    names = set()
+
+    # The NER model used here was trained on affiliation strings rather than
+    # arbitrary text, and as such has a fairly low upper limit of the number
+    # of input tokens (128). At the same time, our input data (GitHub repo
+    # READMEs) will typically be much longer. To overcome this limitation,
+    # we move a sliding window over the input text to obtain multiple shorter
+    # substrings, and each of them is passed to the NER model separately. The
+    # detected organisation names are then concatenated and returned.
+    offset = 0
+    while offset < len(tokens):
+        tokens_chunk = tokens[offset:]
```

Comment:
Should this be `tokens[offset:offset+window_size]` where `window_size=MAX_NER_INPUT_LEN - 8`?

Reply:
It would be `tokens[offset:offset+MAX_NER_INPUT_LEN]`, as the window moves by `MAX_NER_INPUT_LEN - 8`, but is `MAX_NER_INPUT_LEN` wide.

Previously I thought this was not necessary, as the tokenizer removes the tokens above the limit anyway (line 47). But I still modified it like you suggested, it is less confusing and also should use less memory as we copy shorter substrings.

Reply:
Ok that makes sense, thanks for the explanation!